### ISSUE TITLE
Investigate Apache cTAKES™

### ISSUE THREAD

**Issue by MatthewVita** [openemr/openemr#800]
## PROJECT

The Chart Summarization Research project is an effort led by Dr. André Millet. Dr. Millet has put forth an idea of using an algorithm to summarize the most vital parts of a patient's timeline. This research project will look into the technological, regulatory, and verification requirements for the development of such a module.

## ISSUE DESCRIPTION

This research issue involves diving into Apache cTAKES™, an open source natural language processing system for the extraction of information from electronic medical record clinical free-text.

## TASKS

- [x] Determine if this technology is production ready and would be beneficial for the Chart Summarization Research project.
- [x] Put together a simple demo of cTAKES™ using a few de-identified SOAP notes.

**Comment by robertdown** [comment]
I wonder if this would be an area where machine learning could be beneficial. I know pretty much nothing about how it works, and have only started watching a few videos on it... just a thought

**Comment by bradymiller** [comment]
First step may be having a gui that shows listing of all patient forms (without needing to worry about encounters since not clinically relevant). Then could go from there:
http://www.open-emr.org/wiki/index.php/Active_Projects#Ability_to_view_all_notes_from_one_screen

**Comment by MatthewVita** [comment]
> I wonder if this would be an area where machine learning could be beneficial.

My understanding is that cTAKES uses machine learning models under the hood that are trained using clinically annotated golden data. Here are the components that make up cTAKES: https://ctakes.apache.org/components.html (you can see which ones use ML).

> First step may be having a gui that shows listing of all patient forms

My current understanding is that this feature is valuable, but I believe André wants to take it several steps further with something the Open Health Natural Language Processing Consortium calls "Text Summarization". Here's how they define this particular use case "[...] to accelerate the chart review process so that clinical information in multiple reports of the same patient can be extracted and visualized so that clinicians can quickly grasp major past medical history of a patient without performing lengthy chart reviews. [...]"

> without needing to worry about encounters since not clinically relevant

I'm confused, I thought encounters were clinically relevant. This is where my lack of clinical knowledge starts to show :)

**Comment by bradymiller** [comment]
@MatthewVita ,

Agree that what you are proposing is way more than a gui listing of patient forms.

Encounters dates are redundant and really just a mechanism to bundle visits for billing. Clinically, for a physician, these dates can simply be noted by the date of the actual clinical patient forms.

Currently, OpenEMR does not have a gui that simply provides a list of all patient forms(and without distraction of encounter separations) in chronological order that can be easily reviewed by clicking on each note.

In Patient Reports, can select all the notes and then create a report with them in it (and can even search them via a javascript feature that was required for MU2). In Encounter listing, can see the forms bundled with encounters and see them when hover over them.

But, there is still not just a simple gui that lists the forms and allow quick review/view of them. For physicians, being able to do this efficiently (rather than needing to traverse above steps) would be helpful. This could even be a building block towards cTAKES solution(ie. may be able to), or it may not be :)

Note this is my opinion and I've always wondered why none of the practicing physicians that are using OpenEMR have requested this.

**Comment by MatthewVita** [comment]
> Encounters dates are redundant and really just a mechanism to bundle visits for billing. Clinically, for a physician, these dates can simply be noted by the date of the actual clinical patient forms.

I think I see the point of my confusion. I thought you were saying we needn't include the encounter data when running the NLP engine, but it looks like you are saying we don't really need to care about the dates because the forms are all marked by date in a bundle. Okay, makes sense now.

> Currently, OpenEMR does not have a gui that simply provides a list of all patient forms(and without distraction of encounter separations) in chronological order [...] I've always wondered why none of the practicing physicians using OpenEMR have requested this.

So this is the area of interest. Andre stated in his initial email that he would like such a feature, but it really seems like his needs go deeper from a summarization perspective (which made my mind go right to ML/NLP processing, based on his examples). I think the solution here is to have Andre lead a small team of volunteers in our community and see what kind of "chronological patient timeline" information can be gained from cTAKES. If we find that it is simply not feasible from a data quality perspective, we should fall back on the simple "quick review" GUI that you mentioned.

If we get the cTAKES implementation right and provide the user with a great timeline view, it could be a game changer!

**Comment by MatthewVita** [comment]
...advanced docker image is almost ready :)

**Comment by MatthewVita** [comment]
Dr. Millet and I are testing out the solution! Looks like free form encounter data is what cTAKES is good at.

**Comment by MatthewVita** [comment]
As I mentioned in another ticket

> This can be closed. Dr. Millet and I have been talking about this in chat for weeks now. Basically we want to run cTAKES on all non-structured data for a given patient and show the important bits as a "timeline" on the main patient page. The structured data can be summarized there as well.

> Another great usage of the solution is with billing/insurance... basically codes can be determined by cTAKES.

> Not sure how we want to distribute the solution because Docker will be needed and that's not the easiest to set up. Perhaps we can advertise it as an "add on" to OpenEMR in the Wiki and the website with a switch in globals that is defaulted to false.

Once this ticket is done, I will close the project and create an implementation project.

**Comment by bradymiller** [comment]
Agree with offering as an add on and using a globals switch to turn on.

**Comment by girauch** [comment]
Hello- Do you need a tech writer to help you with documentation? I'm entering the field of tech writing, coming from a user experience background. It would be great to get a few projects under my belt to start a portfolio. Let me know! I'm on LinkedIn (https://www.linkedin.com/in/christinefgiraud/) and my website with resume: www.christinefgiraud.com.
